spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/ssl/FileWatcher.java

```diff
@@ -0,0 +1,302 @@
+
+	private final String threadName;
+
+	private final Duration quietPeriod;
+
+	private final Object lifecycleLock = new Object();
+
+	private final Map<WatchKey, List<Registration>> registrations = new ConcurrentHashMap<>();
+
+	private volatile WatchService watchService;
+
+	private Thread thread;
+
+	private boolean running = false;
+
+	FileWatcher(String threadName, Duration quietPeriod) {
+		Assert.notNull(threadName, "threadName must not be null");
+		Assert.notNull(quietPeriod, "quietPeriod must not be null");
+		this.threadName = threadName;
+		this.quietPeriod = quietPeriod;
+	}
+
+	void watch(Set<Path> paths, Callback callback) {
+		Assert.notNull(callback, "callback must not be null");
+		Assert.notNull(paths, "paths must not be null");
+		if (paths.isEmpty()) {
+			return;
+		}
+		startIfNecessary();
+		try {
+			registerWatchables(callback, paths, this.watchService);
+		}
+		catch (IOException ex) {
+			throw new UncheckedIOException("Failed to register paths for watching: " + paths, ex);
+		}
+	}
+
+	void stop() {
+		synchronized (this.lifecycleLock) {
+			if (!this.running) {
+				return;
+			}
+			this.running = false;
+			this.thread.interrupt();
+			try {
+				this.thread.join();
+			}
+			catch (InterruptedException ex) {
+				Thread.currentThread().interrupt();
+			}
+			this.thread = null;
+			this.watchService = null;
+			this.registrations.clear();
+		}
+	}
+
+	private void startIfNecessary() {
+		synchronized (this.lifecycleLock) {
+			if (this.running) {
+				return;
+			}
+			CountDownLatch started = new CountDownLatch(1);
+			this.thread = new Thread(() -> this.threadMain(started));
+			this.thread.setName(this.threadName);
+			this.thread.setDaemon(true);
+			this.thread.setUncaughtExceptionHandler(this::onThreadException);
+			this.running = true;
+			this.thread.start();
+			try {
+				started.await();
+			}
+			catch (InterruptedException ex) {
+				Thread.currentThread().interrupt();
+			}
+		}
+	}
+
+	private void threadMain(CountDownLatch started) {
+		logger.debug("Watch thread started");
+		try (WatchService watcher = FileSystems.getDefault().newWatchService()) {
+			this.watchService = watcher;
+			started.countDown();
+			Map<Registration, List<Change>> accumulatedChanges = new HashMap<>();
+			while (this.running) {
+				try {
+					WatchKey key = watcher.poll(this.quietPeriod.toMillis(), TimeUnit.MILLISECONDS);
+					if (key == null) {
+						// WatchService returned without any changes
+						if (!accumulatedChanges.isEmpty()) {
+							// We have queued changes, that means there were no changes
+							// since the quiet period
+							fireCallback(accumulatedChanges);
+							accumulatedChanges.clear();
+						}
+					}
+					else {
+						accumulateChanges(key, accumulatedChanges);
+					}
+				}
+				catch (InterruptedException ex) {
+					Thread.currentThread().interrupt();
+				}
+			}
+			logger.debug("Watch thread stopped");
+		}
+		catch (IOException ex) {
+			throw new UncheckedIOException(ex);
```

Comment:
If an exception is thrown, this thread dies and stays dead. Should we make this more robust and keep running in case of exceptions?